### DISCUSSION TITLE
About Angle Rate Controller (Observer Design)

### DISCUSSION THREAD

**Discussion by doltae09** [bitcraze/discussions#324] [Q&A] [answered]
Hello, I'm Sejong University's Student in South Korea. In out lab, we use the Crazyflie platform for many research. I am very thankful to your company for allowing us to do a lot of research.

I am designing rollRateNew (𝝓'_𝒏𝒆𝒘) from My method’s estimation data (Observer).  

In this situation, I want to change from the original feedback data (rollRateActual) to My method’s estimation data(Observer).  

How can I change original feedback data in firmware “controller_pid.c” or "attitude_pid_controller.c"??
Can you give me some information the change point?
 
![Question_221018_2](https://user-images.githubusercontent.com/79627957/196342049-2d17b0d5-24fb-4d57-bc44-6a95f6d4c79e.png)

**Comment by krichardsson** [comment, marked as answer]
Hi,
thanks for the kind words!

I think you probably want to dig in somewhere around [here in controller_pid.c](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/controller_pid.c#L122-L127)

It looks like the call to `attitudeControllerCorrectRatePID()` is doing the angle rate updates (when calling `pidUpdate()`) and thus running the "angle rate controller". The result from the angle rate controller is stored in [the static variables `rollOutput`, `pitchOutput` and `yawOutput`](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/attitude_pid_controller.c#L99-L101) that finally are read back using the `attitudeControllerGetActuatorOutput()` function. 

(I think it is fair to say that there is room for improvements for this code...)

I'm not exactly sure where your data lives, but if it is in controller_pid.c I guess you can just add more arguments to `attitudeControllerCorrectRatePID()` to pass your data on. 

I don't know if this answers your question?

**Comment by doltae09** [comment]
Thank you! krichardsson!
I can change the feedback data for Angle Rate feedback data!
